### PR TITLE
Add live model persistence to the frontend

### LINKED ISSUE #71: Live model persistence - initial pass of saving one model per user
Add the ability to save the current state of the model for a given user, and automatically reload the form when they return. As the form is updated, the new model parameters should be automatically saved to the backend.

### LINKED ISSUE #91: URL-encoded state does not keep track of blanked-out fields
**Describe the bug**
To mostly copy from @sprad's comment [here](https://github.com/Recidiviz/covid19-dashboard/pull/90#issuecomment-612642745), when you reload the model page, the "Ages Unknown" current case field is repopulating with the default regardless of whether or not that field was changed.

This behavior holds for any field that has a default value.

**To Reproduce**

1.  Cut and Paste the value in the "Ages Unknown" field into the Ages 0-19 field.

Screen Shot 2020-04-12 at 12 34 24 PM

2.  Reload the page

3.  Notice the "Ages Unknown" field has been repopulated with the default value

Screen Shot 2020-04-12 at 12 32 02 PM

**Expected behavior**
Fields that are blanked-out by the user should remain blank when the page is reloaded.

**Screenshots**

**Desktop (please complete the following information):**

- OS [e.g. Windows 10]: macOS 10.15.4
- Browser [e.g. chrome, edge]: Chrome
- Version [e.g. 22]: 80.0.3987.163

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by itsjon**
## Description of the change

(Builds on unmerged PR https://github.com/Recidiviz/covid19-dashboard/pull/85.)

Updates frontend to autosave model inputs as the user changes them. Per discussion with Andrew, hydrating the state via query parameters will still work and will supersede state saved to Firestore. We also decided it's acceptable that selecting a different state/county will blow away saved state (since local state resets) since this persistence should be updated in a matter of days.

Manual testing performed:

1.  Start with all site storage cleared and no data for user in Firestore
2. Go to the app, without any query params, and sign in
3. Change the state selection and a few population numbers
4. Revisit the app (again, no query params) and verify state retained
5. Copy the full URL (with the state params), modify a param, and load that URL
6. Verify the modified param is respected in the UI
7. Revisit the app (again, no query params) and verify the state set by the modified query param is retained
8. Sign out and sign back in and verify the state is unchanged

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

Closes #71

Closes #91

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Comment by itsjon** [comment]
This includes one commit beyond what's in https://github.com/Recidiviz/covid19-dashboard/pull/85 that updates the data models so there's just one set of inputs per document in the `model_inputs` collection. Based on what I've seen in the multi-facility designs, I think we'll want another collection called something like `models`, which includes subcollections of facilities, which each referencing a single model in the `models` collection. I'm not sure that's right, but I think this version has a better chance of working for future iterations.

**Comment by sprad** [comment]
Hey @itsjon - great work again.

I've checked out the branch locally to play around with it.  Overall, things look great.  I've noticed a bug though.  When I log out and then log back in, the "Ages Unknown" current case field is repopulating with the default regardless of whether or not that field was changed.

Steps to reproduce:

1.  Cut and Paste the value in the "Ages Unknown" field into the Ages 0-19 field.

Screen Shot 2020-04-12 at 12 34 24 PM

2.  Log out and Log Back In

3.  Notice the "Ages Unknown" field has been repopulated with the default value

Screen Shot 2020-04-12 at 12 32 02 PM

**Comment by sprad** [comment]
Seems you don't even have to log out and back in.  You can just refresh the page to experience the bug.

**Comment by itsjon** [comment]
@sprad good catch! Yeah, I think this will be the case for any field that's deleted by the user and has a default value. I'll push up a fix.

**Comment by itsjon** [comment]
Looks like that behavior is also an issue for the URL-based persistence. Let me chat with Andrew about this. I'm not sure if we want to "freeze" the data when it first loads, so the user will definitely see the exact same inputs when they return to the model (even if there is more up-to-data data in the spreadsheet), or specifically only save data the user changes and otherwise continue to use data from the spreadsheet. I'm guessing the former, but I think we should make an explicit decision.

**Comment by sprad** [comment]
I'm going to ask an annoying question so feel free to ignore.

Do we still need to store the state in the query parameters now that Firebase is in place?  I'm assuming this is for link sharing?  Maybe there's another way?  Just food for thought.  Shouldn't block this PR from going in

File: src/impact-dashboard/EpidemicModelContext.tsx [hunk 1]
```diff
@@ -220,20 +215,6 @@ function epidemicModelReducer(
       }
 
       return Object.assign({}, state, updates, { stateInitialized });
-
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/impact-dashboard/EpidemicModelContext.tsx]: Thank you!

**Comment by itsjon** [comment]
> I'm going to ask an annoying question so feel free to ignore.
> 
> Do we still need to store the state in the query parameters now that Firebase is in place? I'm assuming this is for link sharing? Maybe there's another way? Just food for thought. Shouldn't block this PR from going in

Fair question. For now Andrew and I decided to keep it, but I imagine it might not be long before the URL is no longer constantly updated and instead a parameterized URL can only be obtained via a specific link sharing feature in the UI.

**Comment by itsjon** [comment]
OK, calling this done for now. If it looks good to y'all, feel free to merge.

**Pull request merged**